engine/access/rpc/backend/backend_transactions.go

```diff
@@ -1039,3 +1040,292 @@ func (b *backendTransactions) tryGetTransactionResultByIndex(
+) (map[flow.Identifier]string, error) {
+	txResults, err := b.results.ByBlockID(blockID)
+	if err != nil {
+		return nil, rpc.ConvertStorageError(err)
+	}
+
+	results := make(map[flow.Identifier]string)
+	needToFetch := false
+	for _, txResult := range txResults {
+		if txResult.Failed {
+			cacheKey := flow.MakeIDFromFingerPrint(append(blockID[:], txResult.TransactionID[:]...))
+			if value, ok := b.txErrorMessagesCache.Get(cacheKey); ok {
+				results[txResult.TransactionID] = value
+			} else {
+				needToFetch = true
+			}
+		}
+	}
+
+	// all transactions were served from cache or there were no failed transactions
+	if !needToFetch {
+		return results, nil
+	}
+
+	execNodes, err := executionNodesForBlockID(ctx, blockID, b.executionReceipts, b.state, b.log)
+	if err != nil {
+		if IsInsufficientExecutionReceipts(err) {
+			return nil, status.Errorf(codes.NotFound, err.Error())
+		}
+		return nil, rpc.ConvertError(err, "failed to retrieve error message from any execution node", codes.Internal)
+	}
+	req := &execproto.GetTransactionErrorMessagesByBlockIDRequest{
+		BlockId: convert.IdentifierToMessage(blockID),
+	}
+
+	resp, err := b.getTransactionErrorMessagesFromAnyEN(ctx, execNodes, req)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch error message from ENs: %w", err)
+	}
+	result := make(map[flow.Identifier]string, len(resp))
+	for _, value := range resp {
+		cacheKey := flow.MakeIDFromFingerPrint(append(req.BlockId, value.TransactionId...))
+		b.txErrorMessagesCache.Add(cacheKey, value.ErrorMessage)
+		result[convert.MessageToIdentifier(value.TransactionId)] = value.ErrorMessage
+	}
+	return result, nil
+}
+
+// getTransactionErrorMessageFromAnyEN performs an RPC call using available nodes passed as argument.
+func (b *backendTransactions) getTransactionErrorMessageFromAnyEN(
+	ctx context.Context,
+	execNodes flow.IdentityList,
+	req *execproto.GetTransactionErrorMessageRequest,
+) (*execproto.GetTransactionErrorMessageResponse, error) {
+	var errToReturn error
+
+	// if we were passed 0 execution nodes add a specific error
+	if len(execNodes) == 0 {
+		return nil, errors.New("zero execution nodes")
+	}
+
+	var resp *execproto.GetTransactionErrorMessageResponse
+	errToReturn = b.nodeCommunicator.CallAvailableNode(
+		execNodes,
+		func(node *flow.Identity) error {
+			var err error
+			resp, err = b.tryGetTransactionErrorMessageFromEN(ctx, node, req)
+			if err == nil {
+				b.log.Debug().
+					Str("execution_node", node.String()).
+					Hex("block_id", req.GetBlockId()).
+					Hex("transaction_id", req.GetTransactionId()).
+					Msg("Successfully got transaction error message from any node")
+				return nil
+			}
+			return err
+		},
+		nil,
+	)
+
+	// log the errors
+	if errToReturn != nil {
+		b.log.Err(errToReturn).Msg("failed to get transaction error message from execution nodes")
+	}
+
+	return resp, errToReturn
+}
+
+// getTransactionErrorMessageFromAnyEN performs an RPC call using available nodes passed as argument.
```

Comment:
```suggestion
// getTransactionErrorMessageFromAnyEN performs an RPC call using available nodes passed as argument. List of nodes must be non-empty otherwise an error will be returned.
// Expected errors during normal operation:
//   - status.Error - GRPC call failed, some of possible codes are:
//      - codes.NotFound - request cannot be served by EN because of absence of data.
//      - codes.Unavailable - remote node is not unavailable.
```